numpy/core/src/umath/ufunc_object.c

```diff
@@ -3128,9 +3041,9 @@ PyUFunc_Accumulate(PyUFuncObject *ufunc, PyArrayObject *arr, PyArrayObject *out,
     need_outer_iterator = (ndim > 1);
     /* We can't buffer, so must do UPDATEIFCOPY */
     if (!PyArray_ISALIGNED(arr) || (out && !PyArray_ISALIGNED(out)) ||
-            !PyArray_EquivTypes(op_dtypes[0], PyArray_DESCR(arr)) ||
+            !PyArray_EquivTypes(descrs[1], PyArray_DESCR(arr)) ||
             (out &&
-             !PyArray_EquivTypes(op_dtypes[0], PyArray_DESCR(out)))) {
+             !PyArray_EquivTypes(descrs[0], PyArray_DESCR(out)))) {
```

Comment:
The code always assumed that all descriptors are identical (and I added an assert for it).  Using `1` for the operand and `0` for the output seems clearer to me, since that is what is used for the inner-loop at the end.  Arguably, it could also be `descrs[2]` for the output (output `2` and first intput `0` have to be the same for a reduce-like, but `0` and `1` do not have to be the same for reductions, but are here for reduceat/accumulate).